libcudacxx/include/cuda/std/__utility/typeid.h

```diff
@@ -0,0 +1,507 @@
+{
+  struct __pretty_name_end;
+};
+
+// Get the type name from the pretty name by trimming the front and back.
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr __string_view __pretty_nameof_3(__string_view __sv) noexcept
+{
+  return __sv.substr(__sv.find("__pretty_name_begin<") + sizeof("__pretty_name_begin<") - 1,
+                     __sv.find_end(">::__pretty_name_end"));
+}
+
+template <class _Tp>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr __string_view __pretty_nameof_2() noexcept
+{
+#if defined(_CCCL_COMPILER_GCC) && _CCCL_GCC_VERSION < 90000
+  return __pretty_nameof_3(__make_pretty_name<_Tp>(integral_constant<size_t, ~size_t(0)>{}));
+#else
+  return __pretty_nameof_3(__string_view(_CCCL_PRETTY_FUNCTION));
+#endif
+}
+
+template <class _Tp>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr __string_view __pretty_nameof() noexcept
+{
+  return __pretty_nameof_2<typename __pretty_name_begin<_Tp>::__pretty_name_end>();
+}
+
+// A quick smoke test to ensure that the pretty name extraction is working.
+static_assert(3 == __string_view("int").size(), "3 == __string_view(int).size()");
+static_assert(3 == __pretty_nameof<int>().size(), "3 == __pretty_nameof<int>().size()");
+static_assert(__pretty_nameof<int>() == __string_view("int"), "__pretty_nameof<int>() == __string_view(\"int\")");
+static_assert(__pretty_nameof<float>() < __pretty_nameof<int>(), "__pretty_nameof<float>() < __pretty_nameof<int>()");
+
+/// @brief A minimal implementation of `std::type_info` for platforms that do
+/// not support RTTI.
+struct __type_info
+{
+  __type_info()                   = delete;
+  __type_info(__type_info const&) = delete;
+
+  _LIBCUDACXX_HIDE_FROM_ABI explicit constexpr __type_info(__string_view __name) noexcept
+      : __name_(__name)
+  {}
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr char const* name() const noexcept
+  {
+    return __name_.begin();
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr __string_view __name_view() const noexcept
+  {
+    return __name_;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr bool before(const __type_info& __other) const noexcept
+  {
+    return __name_ < __other.__name_;
+  }
+
+  // Not yet implemented:
+  // _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr size_t hash_code() const noexcept
+  // {
+  //   return ;
+  // }
+
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator==(const __type_info& __lhs, const __type_info& __rhs) noexcept
+  {
+    return &__lhs == &__rhs || __lhs.__name_ == __rhs.__name_;
+  }
+
+#if _CCCL_STD_VER <= 2017
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator!=(const __type_info& __lhs, const __type_info& __rhs) noexcept
+  {
+    return !(__lhs == __rhs);
+  }
+#endif
+
+private:
+  __string_view __name_;
+};
+
+#if !defined(_CCCL_NO_INLINE_VARIABLES)
+
+template <class _Tp>
+inline constexpr __type_info __typeid{__pretty_nameof<_Tp>()};
```

Comment:
We should always try and qualify functions because _Tp could pullin other functions via ADL
```suggestion
inline constexpr __type_info __typeid{_CUDA_VSTD::__pretty_nameof<_Tp>()};
```

Reply:
Only function arguments cause argument dependent lookup. No arguments == no ADL.